### DISCUSSION TITLE
Converting DateTime from one timezone to another

### DISCUSSION THREAD

**Discussion by monojk** [bxparks/AceTime#61] [Q&A]
What would be an easy way to convert a given Date Time string, say "10.04.2019 00:30:00" from Berlin TZ to London TZ, also as a string with the same format?
I know I can use `berlinTime.timeOffset().toMinutes()` , `epochSeconds` , `berlinTime = ZonedDateTime::forEpochSeconds(epochSeconds, berlinTz)` and `berlinTime.printTo()`
I currently struggling getting the result formatted into a string.

**Comment by bxparks** [comment]
Hi, If your date string format is not ISO8601 (i.e. yyyy-mm-dd[T, ]hh:mm:ss+/-hh:mm), I'm afraid you will have to write your own parser and formatter using the `ZonedDateTime::year(), month(), day(), hour(), minute(), second()` methods.

There are so many different date formats that I did not want to get into the business of supporting all those different formats. Especially considering how much extra flash memory such functionality would consume. Most of the big date/time libraries targeting desktop environments use formatting specifications using various "%" notations. (Go Lang is a bit weird, and some people like it.) But if you want to build such a date formatter library on top of AceTime, you are totally welcome to do so.

**Reply by bxparks** [reply]
One thing that might help you build your custom string is my `PrintStr<N>` and `PrintStrN` helper classes. See
https://github.com/bxparks/AceCommon/tree/develop/src/print_str

**Comment by monojk** [comment]
Thanks for your hints. I notices that I can query `berlinTime.year();` etc. from the result. I can then format the resulting dateTime string. I also use `strptime` to parse the input dateTime string. With that I have working code.

**Reply by bxparks** [reply]
Yes, if you are on an Arduino platform that has a `strptime()` and `strftime()` (ESP8266 I assume?), then you can create a `struct tm`, and do the conversion using those. One thing I am curious about is how much flash memory something like `strptime()` and `strftime()` brings into the binary. 

I suppose I could add a conversion function to create a `struct tm` if that is available on a platform, but I don't really want to. I want to avoid the old `time.h` functions as much as possible, because they are not available on all Arduino platforms. They also use crusty APIs from 50 years ago which are incredibly confusing, difficult to use, with lots of gotchas (e.g. January == 0; year begins at 1900; can't handle negative DST offsets; uses non-threadsafe static buffers; can't handle multiple timezones at the same time; function names are hard to remember; many useful functions are non-standard, e.g. timegm()).

See http://www.catb.org/esr/time-programming/ for a good overview. Here's a nice quote fpr `strptime()`: "Test code that uses strptime(3) carefully on real data. Implementations are bug-prone." Or this one for `strftime()`: "Some of the format specifiers need to be used with caution; older implementations may not support all of them."
